tensorflow_addons/image/filters.py

```diff
@@ -142,80 +128,80 @@ def mean_filter2d(image,
-        if row < filter_shapex or col < filter_shapey:
+        image = tf.convert_to_tensor(image, name="image")
+
+        rank = image.shape.rank
+        if rank != 3 and rank != 4:
+            raise ValueError("image should be either 3 or 4-dimensional.")
+
+        if padding not in ["REFLECT", "CONSTANT", "SYMMETRIC"]:
             raise ValueError(
-                'Number of Pixels in each dimension of the image should be \
-                more than the filter size. Got filter_shape (%sx' %
-                filter_shape[0] + '%s).' % filter_shape[1] +
-                ' Image Shape (%s)' % image.shape)
-        if filter_shapex % 2 == 0 or filter_shapey % 2 == 0:
-            raise ValueError('Filter size should be odd. Got filter_shape '
-                             '(%sx%s)' % (filter_shape[0], filter_shape[1]))
-        image = tf.cast(image, tf.float32)
-        tf_i = tf.reshape(image, [row * col * ch])
-        ma = tf.math.reduce_max(tf_i)
-        image = _normalize(image, ma)
-
-        # k and l is the Zero-padding size
-
-        listi = []
-        for a in range(ch):
-            img = image[:, :, a:a + 1]
-            img = tf.reshape(img, [1, row, col, 1])
-            slic = tf.image.extract_patches(
-                img, [1, filter_shapex, filter_shapey, 1], [1, 1, 1, 1],
-                [1, 1, 1, 1],
-                padding='SAME')
-            mid = int(filter_shapex * filter_shapey / 2 + 1)
-            top = tf.nn.top_k(slic, mid, sorted=True)
-            li = tf.slice(top[0], [0, 0, 0, mid - 1], [-1, -1, -1, 1])
-            li = tf.reshape(li, [row, col, 1])
-            listi.append(li)
-        y = tf.concat(listi[0], 2)
-
-        for i in range(len(listi) - 1):
-            y = tf.concat([y, listi[i + 1]], 2)
-
-        y *= 255
-        y = tf.cast(y, tf.int32)
-
-        return y
+                "padding should be one of \"REFLECT\", \"CONSTANT\", or "
+                "\"SYMMETRIC\".")
+
+        filter_shape = keras_utils.conv_utils.normalize_tuple(
+            filter_shape, 2, "filter_shape")
+
+        # Expand to a 4-D tensor
+        if rank == 3:
+            image = tf.expand_dims(image, axis=0)
+
+        # Explicitly pad the image
+        image = _pad(
+            image, filter_shape, mode=padding, constant_values=constant_values)
+
+        floor = (filter_shape[0] * filter_shape[1] + 1) // 2
+        ceil = (filter_shape[0] * filter_shape[1]) // 2 + 1
+
+        def _median_filter2d_single_channel(x):
+            x = tf.expand_dims(x, axis=-1)
+            patches = tf.image.extract_patches(
+                x,
+                sizes=[1, filter_shape[0], filter_shape[1], 1],
+                strides=[1, 1, 1, 1],
+                rates=[1, 1, 1, 1],
+                padding="VALID")
+            top = tf.nn.top_k(patches, k=ceil).values
+            median = (top[:, :, :, floor - 1] + top[:, :, :, ceil - 1]) / 2
+            return tf.dtypes.cast(median, x.dtype)
```

Comment:
Do we need to document the behavior? Say, return 6, rather than 6.5, for [5, 6, 7, 8]

Reply:
Sure, it makes sense